firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/core/CrashlyticsController.java

```diff
@@ -1805,26 +1809,84 @@ private void sendSessionReports(AppSettingsData appSettings, boolean dataCollect
+
+      if (eventLatch.await(2000, TimeUnit.MILLISECONDS)) {
+        Logger.getLogger().d(Logger.TAG, "Crash event callback received from FA listener.");
+      } else {
+        Logger.getLogger()
+            .d(
+                Logger.TAG,
+                "Timeout exceeded while awaiting crash event callback from FA listener.");
+      }
     }
 
-    if (analyticsConnector != null) {
-      Logger.getLogger().d(Logger.TAG, "Logging Crashlytics event to Firebase");
-      final Bundle params = new Bundle();
-      params.putInt(FIREBASE_CRASH_TYPE, FIREBASE_CRASH_TYPE_FATAL);
-      params.putLong(FIREBASE_TIMESTAMP, timestamp);
-      analyticsConnector.logEvent(
-          FIREBASE_ANALYTICS_ORIGIN_CRASHLYTICS, FIREBASE_APPLICATION_EXCEPTION, params);
-    } else {
-      Logger.getLogger()
-          .d(Logger.TAG, "Skipping logging Crashlytics event to Firebase, no Firebase Analytics");
+    @Override
+    public void onCrashOriginEvent(int id, Bundle extras) {
+      String eventName = extras.getString(AnalyticsConnectorReceiver.EVENT_NAME_KEY);
+      if (AnalyticsConnectorReceiver.APP_EXCEPTION_EVENT_NAME.equals(eventName)) {
+        eventLatch.countDown();
+      }
     }
   }
 
+  /**
+   * Send an App Exception event to Firebase Analytics. FA records the event asynchronously, so this
+   * method returns a Task in case the caller wants to verify that the event was recorded by FA and
+   * will not be lost.
+   */
+  private Task<Void> recordFatalFirebaseEvent(long timestamp) {
+    final ThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1);
+    return Tasks.call(
+        executor,
+        new Callable<Void>() {
+          @Override
+          public Void call() throws Exception {
+            if (firebaseCrashExists()) {
+              Logger.getLogger()
+                  .d(
+                      Logger.TAG,
+                      "Skipping logging Crashlytics event to Firebase, FirebaseCrash exists");
+              return null;
+            }
+            if (analyticsConnector == null) {
+              Logger.getLogger()
+                  .d(
+                      Logger.TAG,
+                      "Skipping logging Crashlytics event to Firebase, no Firebase Analytics");
+              return null;
+            }
+            final BlockingCrashEventListener blockingListener = new BlockingCrashEventListener();
+            analyticsReceiver.setCrashlyticsOriginEventListener(blockingListener);
+
+            Logger.getLogger().d(Logger.TAG, "Logging Crashlytics event to Firebase");
+            final Bundle params = new Bundle();
+            params.putInt(FIREBASE_CRASH_TYPE, FIREBASE_CRASH_TYPE_FATAL);
+            params.putLong(FIREBASE_TIMESTAMP, timestamp);
+
+            analyticsConnector.logEvent(
+                FIREBASE_ANALYTICS_ORIGIN_CRASHLYTICS, FIREBASE_APPLICATION_EXCEPTION, params);
+
+            blockingListener.awaitEvent();
```

Comment:
If this returned a task (as mentioned above), you could remove the extra `Tasks.call` above and turn this method (`recordFatalFirebaseEvent`) into a synchronous call (which still returns this task), then just chain the return value in the uncaught exception handler the same way.

Reply:
We both like this idea and will do this in a subsequent refactor, since it is going to require more extensive validation than we have time for before the next release. It mostly improves readability and not performance.